oauth_dropins/indieauth.py

```diff
@@ -149,35 +205,61 @@ def button_html(cls, *args, **kwargs):
 
 
 class Callback(views.Callback):
-  """The callback view from the IndieAuth request. POSTs back to the
-  auth endpoint to verify the authentication code."""
+  """The callback view from the IndieAuth request. Performs an Authorization
+  Code grant to verify the code."""
   def dispatch_request(self):
     code = request.values['code']
     state = util.decode_oauth_state(request.values['state'])
 
-    endpoint = state.get('endpoint')
+    token_endpoint = state.get('token_endpoint')
     me = state.get('me')
-    if not endpoint or not me:
-      flask_util.error("invalid state parameter")
-
-    state = state.get('state') or ''
-    validate_resp = util.requests_post(endpoint, data={
-      'me': me,
-      'client_id': INDIEAUTH_CLIENT_ID,
-      'code': code,
-      'redirect_uri': request.base_url,
-      'state': state,
-    })
-
-    if validate_resp.status_code // 100 == 2:
-      data = util.sniff_json_or_form_encoded(validate_resp.text)
-      if data.get('me'):
-        verified = data.get('me')
-        user_json = build_user_json(verified)
-        indie_auth = IndieAuth(id=verified, user_json=json_dumps(user_json))
-        indie_auth.put()
-        return self.finish(indie_auth, state=state)
+    if token_endpoint:
+      # TODO: validate that the `iss` matches the value that is retrieved from the IndieAuth Server Metadata https://indieauth.spec.indieweb.org/#indieauth-server-metadata
+      code_verifier = state.get('code_verifier') or ''
+      validate_resp = util.requests_post(token_endpoint, data={
+        'grant_type': 'authorization_code',
+        'client_id': INDIEAUTH_CLIENT_ID,
+        'code': code,
+        'redirect_uri': request.base_url,
+        'code_verifier': code_verifier,
+      })
+
+      if validate_resp.status_code // 100 == 2:
```

Comment:
looks like this onward is the same in both methods, mind pulling it out to the bottom so it's unified?